controlplane/pkg/sidecars/nsm-monitor.go

```diff
@@ -0,0 +1,238 @@
+		stop:        make(chan bool),
+	}
+}
+
+func (c *nsmMonitorApp) beginMonitoring() {
+	for {
+		var configuration *common.NSConfiguration
+		if c.helper != nil {
+			configuration = c.helper.GetConfiguration()
+		}
+		nsmClient, err := client.NewNSMClient(context.Background(), configuration)
+		if err != nil {
+			logrus.Fatalf(nsmMonitorLogWithParamFormat, "Unable to create the NSM client", err)
+
+			c.waitRetry()
+			continue
+		}
+
+		logrus.Infof(nsmMonitorLogFormat, "connection to NSM established")
+
+		monitorClient, err := local.NewMonitorClient(nsmClient.NsmConnection.GrpcClient)
+		if err != nil {
+			logrus.Fatalf(nsmMonitorLogWithParamFormat, "failed to start monitor client", err)
+
+			c.waitRetry()
+			continue
+		}
+		defer monitorClient.Close()
+
+		initRecieved := false
+		recovery := false
+
+		for {
+			if initRecieved && !recovery {
+				// Performing recovery if required.
+				if c.helper != nil {
+					c.helper.Connected(c.connections)
+				}
+				// Since NSMD will setup public socket only when all connections will be ok, we need to perform request only on ones it loose.
+				if c.performRecovery(nsmClient) {
+					// since we not recovered, we will continue after delay
+					continue
+				} else {
+					recovery = true
+				}
+			}
+			select {
+			case err = <-monitorClient.ErrorChannel():
+				logrus.Fatalf(nsmMonitorLogWithParamFormat, "NSM die, re-connecting", err)
+				for _, c := range c.connections {
+					c.State = connection.State_DOWN // Mark all as down.
+				}
+				continue
+			case event := <-monitorClient.EventChannel():
+				if event.EventType() == monitor.EventTypeInitialStateTransfer {
+					logrus.Infof(nsmMonitorLogFormat, "Monitor started")
+					initRecieved = true
+				}
+
+				for _, entity := range event.Entities() {
+					switch event.EventType() {
+					case monitor.EventTypeInitialStateTransfer, monitor.EventTypeUpdate:
+						c.updateConnection(entity)
+					case monitor.EventTypeDelete:
+						logrus.Infof(nsmMonitorLogFormat, "Connection closed")
+						return
+					}
+				}
+			case <-c.stop:
+				if c.helper != nil {
+					c.helper.Stopped()
+					logrus.Infof(nsmMonitorLogFormat, "Processing stop")
+					return
+				}
+			}
+		}
+	}
+}
+
+func (c *nsmMonitorApp) updateConnection(entity monitor.Entity) {
+	conn, ok := entity.(*connection.Connection)
+	// update connections
+	if ok {
+		if existingConn, exists := c.connections[conn.Id]; exists {
+			logrus.Infof(nsmMonitorLogWithParamFormat, "Connection updated", fmt.Sprintf("%v %v", existingConn, conn))
+		} else {
+			logrus.Infof(nsmMonitorLogWithParamFormat, "Initial connection accepted", conn)
+		}
+		c.connections[conn.Id] = conn
+	}
+}
+
+func (c *nsmMonitorApp) waitRetry() {
+	logrus.Errorf(nsmMonitorLogWithParamFormat, "Retry delay %v sec", nsmMonitorRetryDelay)
+	<-time.After(nsmMonitorRetryDelay * time.Second)
+}
+
+func (c *nsmMonitorApp) performRecovery(nsmClient *client.NsmClient) bool {
```

Comment:
Contract of that function is not obvious enough. It returns `false` if everything is fine and `true` if we need retry. Maybe makes sense to do vise versa?